ClusterObserver/Observers/ClusterObserver.cs

```diff
@@ -0,0 +1,171 @@
+                {
+                    // If in Warning and you are not sending Warning state reports, then end here...
+                    if (!emitWarningDetails && clusterHealth.AggregatedHealthState == HealthState.Warning)
+                    {
+                        return;
+                    }
+
+                    var unhealthyEvaluations = clusterHealth.UnhealthyEvaluations;
+
+                    foreach (var evaluation in unhealthyEvaluations)
+                    {
+                        token.ThrowIfCancellationRequested();
+
+                        telemetryDescription += $"{Enum.GetName(typeof(HealthEvaluationKind), evaluation.Kind)} - {evaluation.AggregatedHealthState}: {evaluation.Description}{Environment.NewLine}";
+
+                        // Application in error/warning?...
+                        foreach (var app in clusterHealth.ApplicationHealthStates)
+                        {
+                            if (app.AggregatedHealthState == HealthState.Ok
+                                || (emitWarningDetails && app.AggregatedHealthState != HealthState.Warning))
+                            {
+                                continue;
+                            }
+
+                            telemetryDescription += $"Application in Error or Warning: {app.ApplicationName}{Environment.NewLine}";
+                        }
+                    }
+                }
+
+                // Track current health state for use in next run...
+                this.ClusterHealthState = clusterHealth.AggregatedHealthState;
+
+                // This means there is no cluster health state data to emit...
+                if (string.IsNullOrEmpty(telemetryDescription))
+                {
+                    return;
+                }
+
+                // Telemetry...
+                await this.ObserverTelemetryClient?.ReportHealthAsync(
+                        HealthScope.Cluster,
+                        "AggregatedClusterHealth",
+                        clusterHealth.AggregatedHealthState,
+                        telemetryDescription,
+                        this.ObserverName,
+                        this.Token);
+            }
+            catch (ArgumentException ae) 
+            { 
+                this.ObserverLogger.LogError(
+                    "Unable to determine cluster health:{0}{1}",
+                    Environment.NewLine,
+                    ae.ToString()); 
+            }
+            catch (FabricException fe) 
+            { 
+                this.ObserverLogger.LogError(
+                    "Unable to determine cluster health:{0}{1}",
+                    Environment.NewLine,
+                    fe.ToString()); 
+            }
+            catch (TimeoutException te) 
+            { 
+                this.ObserverLogger.LogError(
+                    "Unable to determine cluster health:{0}{1}",
+                    Environment.NewLine,
+                    te.ToString()); 
+            }
```

Comment:
Why not catch it all in Exception?

Reply:
I only want to throw unhandled exceptions. So, either way I would need to filter by type. If some Fabric or TimeoutException happens, I don't want to throw it after logging. It just means for some exceptional reason cluster health state couldn't be determined. Of course, it may make sense to send that info through the telemetry pipeline in addition to logging on the local machine. Thanks for pointing this out.

Reply:
I decided to just catch, log/emit telemetry, throw. This exception is handled by ObserverManager, in that the FCO process will not be taken down. Only unhandled exceptions via ObserverManager code can take FO or FCO down, by design. A failing observer should not take down the service. In this case, the bug should be fixed and the service redeployed, versus crashing FCO and having SF continually restart it. (See, no ellipses!)